rfcs/0001-messages.md

```diff
@@ -0,0 +1,423 @@
+/// each party’s share of the secret. The actual secret is `f(0)` and the party with
+/// ID `i` will be given a share with value `f(i)`.
+/// Since a DKG may be implemented in the future, we recommend that the ID `0` be declared invalid."
+/// https://raw.githubusercontent.com/ZcashFoundation/redjubjub/main/zcash-frost-audit-report-20210323.pdf#d
+enum ParticipantId {
+    /// A serialized participant ID for a signer.
+    ///
+    /// Must be less than or equal to `MAX_SIGNER_PARTICIPANT_ID`.
+    Signer(u64),
+    /// The fixed participant ID for the dealer.
+    Dealer,
+    /// The fixed participant ID for the aggregator.
+    Aggregator,
+}
+
+/// The fixed participant ID for the dealer.
+const DEALER_PARTICIPANT_ID: u64 = u64::MAX - 1;
+
+/// The fixed participant ID for the aggregator.
+const AGGREGATOR_PARTICIPANT_ID: u64 = u64::MAX;
+
+/// The maximum `ParticipantId::Signer` in this serialization format.
+///
+/// We reserve two participant IDs for the dealer and aggregator.
+const MAX_SIGNER_PARTICIPANT_ID: u64 = u64::MAX - 2;
+```
+
+### Payloads
+
+Each payload defines a new message:
+
+```rust
+/// The data required to serialize `frost::SharePackage`.
+///
+/// The dealer sends this message to each signer for this round.
+/// With this, the signer should be able to build a `SharePackage` and use
+/// the `sign()` function.
+///
+/// Note: `frost::SharePackage.public` can be calculated from `secret_share`.
+struct messages::SharePackage {
+    /// The public signing key that represents the entire group:
+    /// `frost::SharePackage.group_public`.
+    group_public: VerificationKey<SpendAuth>,
+    /// This participant's secret key share: `frost::SharePackage.share.value`.
+    secret_share: frost::Secret,
+    /// The commitments to the coefficients for our secret polynomial _f_,
+    /// used to generate participants' key shares. Participants use these to perform
+    /// verifiable secret sharing.
+    share_commitment: Vec<frost::Commitment>,
+}
+
+/// The data required to serialize `frost::SigningCommitments`.
+///
+/// Each signer must send this message to the aggregator.
+/// A signing commitment from the first round of the signing protocol.
+struct messages::SigningCommitments {
+    /// The hiding point: `frost::SigningCommitments.hiding`
+    hiding: frost::Commitment,
+    /// The binding point: `frost::SigningCommitments.binding`
+    binding: frost::Commitment,
+}
+
+/// The data required to serialize `frost::SigningPackage`.
+///
+/// The aggregator decides what message is going to be signed and
+/// sends it to each signer with all the commitments collected.
+struct messages::SigningPackage {
+    /// The collected commitments for each signer as a hashmap of
+    /// unique participant identifiers: `frost::SigningPackage.signing_commitments`
+    ///
+    /// Signing packages that contain duplicate or missing `ParticipantId`s are invalid.
+    signing_commitments: HashMap<ParticipantId, SigningCommitments>,
```

Comment:
I am considering the use of a `BTreeMap` here instead of a `HashMap` to guanrantee the order in the serialization:

https://github.com/bincode-org/bincode/issues/230

Unsure if we should change the spec or if it is just an implementation detail.

Reply:
I think it would be better to change the spec.
```suggestion
    /// The collected commitments for each signer as a hashmap of
    /// unique participant identifiers: `frost::SigningPackage.signing_commitments`
    ///
    /// Signing packages that contain duplicate or missing `ParticipantId`s are invalid.
    /// `ParticipantId`s must be serialized in ascending numeric order.
    signing_commitments: HashMap<ParticipantId, SigningCommitments>,
```

Reply:
Good catch!